### ISSUE TITLE
Issue with bufferutil during installation

### ISSUE THREAD

**Issue by Overdrivr** [strongloop/strongloop#256]
I am having an issue with bufferutil during installation of strongloop.
Windows 7, npm  3.3.4, node 4.0.1 and a clean install of the Windows SDK for the toolchains
I'm calling 

``` bash
npm install strongloop -g
```

And installation fails on bufferutil package

``` log
128111 info build C:\Users\B48923\AppData\Roaming\npm\node_modules\strongloop\node_modules\bufferutil
128112 info lifecycle bufferutil@1.2.1~preinstall: bufferutil@1.2.1
128113 silly lifecycle bufferutil@1.2.1~preinstall: no script for preinstall, continuing
128114 info linkStuff bufferutil@1.2.1
128115 silly linkStuff bufferutil@1.2.1 has C:\Users\B48923\AppData\Roaming\npm\node_modules\strongloop\node_modules as its parent node_modules
128116 verbose linkBins bufferutil@1.2.1
128117 verbose linkMans bufferutil@1.2.1
128118 verbose rebuildBundles bufferutil@1.2.1
128119 info lifecycle bufferutil@1.2.1~install: bufferutil@1.2.1
128120 verbose lifecycle bufferutil@1.2.1~install: unsafe-perm in lifecycle true
128121 verbose lifecycle bufferutil@1.2.1~install: PATH: C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\bin\node-gyp-bin;C:\Users\B48923\AppData\Roaming\npm\node_modules\strongloop\node_modules\bufferutil\node_modules\.bin;C:\Users\B48923\AppData\Roaming\npm\node_modules\strongloop\node_modules\.bin;C:\Users\B48923\AppData\Roaming\npm\node_modules\.bin;C:\windows\Microsoft.NET\Framework64\v4.0.30319;C:\windows\Microsoft.NET\Framework\v4.0.30319;C:\windows\Microsoft.NET\Framework64\v3.5;C:\windows\Microsoft.NET\Framework\v3.5;;C:\Program Files (x86)\Microsoft Visual Studio 10.0\Common7\IDE;C:\Program Files (x86)\Microsoft Visual Studio 10.0\Common7\Tools;;C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin\amd64;C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin\VCPackages;;C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin\NETFX 4.0 Tools\x64;C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin\x64;C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin;;C:\Python27\;C:\Python27\Scripts;C:\windows\system32;C:\windows;C:\windows\System32\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\QuickTime\QTSystem\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;C:\Program Files (x86)\CodeBlocks\MinGW\bin;C:\Program Files (x86)\CMake\bin;C:\Program Files\WIDCOMM\Bluetooth Software\;C:\Program Files\WIDCOMM\Bluetooth Software\syswow64;C:\Program Files\nodejs\;C:\Program Files\Microsoft SQL Server\120\Tools\Binn\;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files\Microsoft Windows Performance Toolkit\;C:\Program Files\Intel\WiFi\bin\;C:\Program Files\Common Files\Intel\WirelessCommon\;C:\Program Files\Java\jdk1.7.0_79\bin;C:\Users\B48923\AppData\Local\atom\bin;C:\Freescale\KDS_3.0.0\toolchain\bin;C:\Ninja;C:\Program Files (x86)\CMake\bin;C:\Users\B48923\AppData\Roaming\npm;C:\Program Files\MySQL\MySQL Server 5.6\bin
128122 verbose lifecycle bufferutil@1.2.1~install: CWD: C:\Users\B48923\AppData\Roaming\npm\node_modules\strongloop\node_modules\bufferutil
128123 silly lifecycle bufferutil@1.2.1~install: Args: [ '/d /s /c', 'node-gyp rebuild' ]
128124 silly lifecycle bufferutil@1.2.1~install: Returned: code: 1  signal: null
128125 info lifecycle bufferutil@1.2.1~install: Failed to exec install script
128126 verbose unlock done using C:\Users\B48923\AppData\Roaming\npm-cache\_locks\staging-6727808926286ffe.lock for C:\Users\B48923\AppData\Roaming\npm\node_modules\.staging
128127 silly rollbackFailedOptional Starting
128128 silly rollbackFailedOptional Finishing
128129 silly runTopLevelLifecycles Starting
128130 silly runTopLevelLifecycles Finishing
128131 silly install printInstalled
128132 verbose stack Error: bufferutil@1.2.1 install: `node-gyp rebuild`
128132 verbose stack Exit status 1
128132 verbose stack     at EventEmitter.<anonymous> (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\lib\utils\lifecycle.js:233:16)
128132 verbose stack     at emitTwo (events.js:87:13)
128132 verbose stack     at EventEmitter.emit (events.js:172:7)
128132 verbose stack     at ChildProcess.<anonymous> (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\lib\utils\spawn.js:24:14)
128132 verbose stack     at emitTwo (events.js:87:13)
128132 verbose stack     at ChildProcess.emit (events.js:172:7)
128132 verbose stack     at maybeClose (internal/child_process.js:817:16)
128132 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
128133 verbose pkgid bufferutil@1.2.1
128134 verbose cwd C:\Program Files\Microsoft SDKs\Windows\v7.1
128135 error Windows_NT 6.1.7601
128136 error argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\B48923\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "strongloop" "-g"
128137 error node v4.1.0
128138 error npm  v3.3.4
128139 error code ELIFECYCLE
128140 error bufferutil@1.2.1 install: `node-gyp rebuild`
128140 error Exit status 1
128141 error Failed at the bufferutil@1.2.1 install script 'node-gyp rebuild'.
128141 error This is most likely a problem with the bufferutil package,
128141 error not with npm itself.
128141 error Tell the author that this fails on your system:
128141 error     node-gyp rebuild
128141 error You can get their info via:
128141 error     npm owner ls bufferutil
128141 error There is likely additional logging output above.
128142 verbose exit [ 1, true ]
```

Do you have an idea of what is going on ? I'm not sure if it's related to strongloop, node-gyp or bufferutil itself

**Comment by sam-github** [comment]
`bufferutil` is an optional dependency of `ws`, it should not block install. And yes, it is having issues on recent node versions, not sure what the issue you specifically are having is, the output above just says `node-gyp rebuild` returned `1` (a failure).

Can you include a link to a gist (http://gist.github.com) of the entire npm-debug.log file?

**Comment by Overdrivr** [comment]
Thank you for the support,

Here is the complete log file
https://gist.github.com/Overdrivr/70e76d309f7321557df4

**Comment by rmg** [comment]
This will be addressed by the work for #251.

**Comment by rmg** [comment]
Sorry, I spoke before reading all the details of the issue.

This is actually a duplicate of #253, which is ultimately caused by npm/npm#9643.

**TL;DR** there's a bug in npm@3 that causes deeply nested `optionalDependencies` to be treated as mandatory by `npm rebuild` which is what npm@3 does as part of its multi-phase `npm install`.

@Overdrivr can you try downgrading with `npm install -g npm@2` and verifying that this works for you with npm@2?

**Comment by Overdrivr** [comment]
I tried downgrading with your command but that failed. npm refuses to install itself as a dependency. Is there a flag to add to the command ?

```
0 info it worked if it ends with ok
1 verbose cli [ 'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Users\\B48923\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'install',
1 verbose cli   'npm@2',
1 verbose cli   '-g' ]
2 info using npm@3.3.4
3 info using node@v4.1.0
4 silly loadCurrentTree Starting
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 silly fetchPackageMetaData npm@2
8 silly fetchNamedPackageData npm
9 silly mapToRegistry name npm
10 silly mapToRegistry using default registry
11 silly mapToRegistry registry https://registry.npmjs.org/
12 silly mapToRegistry uri https://registry.npmjs.org/npm
13 verbose request uri https://registry.npmjs.org/npm
14 verbose request no auth needed
15 info attempt registry request try #1 at 3:13:45 AM
16 verbose request id 6938d606528834c4
17 verbose etag "9B0WON5H2MDSHUWTOAVLQ6YV"
18 http request GET https://registry.npmjs.org/npm
19 http 304 https://registry.npmjs.org/npm
20 verbose headers { date: 'Tue, 29 Sep 2015 01:13:45 GMT',
20 verbose headers   via: '1.1 varnish',
20 verbose headers   'cache-control': 'max-age=60',
20 verbose headers   etag: '"9B0WON5H2MDSHUWTOAVLQ6YV"',
20 verbose headers   age: '26',
20 verbose headers   connection: 'keep-alive',
20 verbose headers   'x-served-by': 'cache-sjc3125-SJC',
20 verbose headers   'x-cache': 'HIT',
20 verbose headers   'x-cache-hits': '1',
20 verbose headers   'x-timer': 'S1443489225.873668,VS0,VE17',
20 verbose headers   vary: 'Accept' }
21 silly get cb [ 304,
21 silly get   { date: 'Tue, 29 Sep 2015 01:13:45 GMT',
21 silly get     via: '1.1 varnish',
21 silly get     'cache-control': 'max-age=60',
21 silly get     etag: '"9B0WON5H2MDSHUWTOAVLQ6YV"',
21 silly get     age: '26',
21 silly get     connection: 'keep-alive',
21 silly get     'x-served-by': 'cache-sjc3125-SJC',
21 silly get     'x-cache': 'HIT',
21 silly get     'x-cache-hits': '1',
21 silly get     'x-timer': 'S1443489225.873668,VS0,VE17',
21 silly get     vary: 'Accept' } ]
22 verbose etag https://registry.npmjs.org/npm from cache
23 verbose get saving npm to C:\Users\B48923\AppData\Roaming\npm-cache\registry.npmjs.org\npm\.cache.json
24 silly install normalizeTree
25 silly loadCurrentTree Finishing
26 silly loadIdealTree Starting
27 silly install loadIdealTree
28 silly cloneCurrentTree Starting
29 silly install cloneCurrentTreeToIdealTree
30 silly cloneCurrentTree Finishing
31 silly loadShrinkwrap Starting
32 silly install loadShrinkwrap
33 silly loadShrinkwrap Finishing
34 silly loadAllDepsIntoIdealTree Starting
35 silly install loadAllDepsIntoIdealTree
36 silly rollbackFailedOptional Starting
37 silly rollbackFailedOptional Finishing
38 silly runTopLevelLifecycles Starting
39 silly runTopLevelLifecycles Finishing
40 silly install printInstalled
41 verbose stack Error: Refusing to install npm as a dependency of itself
41 verbose stack     at checkSelf (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\lib\install\validate-args.js:40:14)
41 verbose stack     at Array.<anonymous> (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\bind-actor.js:15:8)
41 verbose stack     at LOOP (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:15:14)
41 verbose stack     at chain (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\chain.js:20:5)
41 verbose stack     at C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\lib\install\validate-args.js:15:5
41 verbose stack     at C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\async-map.js:52:35
41 verbose stack     at Array.forEach (native)
41 verbose stack     at C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\async-map.js:52:11
41 verbose stack     at Array.forEach (native)
41 verbose stack     at asyncMap (C:\Users\B48923\AppData\Roaming\npm\node_modules\npm\node_modules\slide\lib\async-map.js:51:8)
42 verbose cwd C:\Users\B48923
43 error Windows_NT 6.1.7601
44 error argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\B48923\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "install" "npm@2" "-g"
45 error node v4.1.0
46 error npm  v3.3.4
47 error code ENOSELF
48 error Refusing to install npm as a dependency of itself
49 error If you need help, you may report this error at:
49 error     <https://github.com/npm/npm/issues>
50 verbose exit [ 1, true ]

```

**Comment by rmg** [comment]
@Overdrivr that's an interesting bug.. It must be npm@3 specific, otherwise I imagine you would have gotten the same error when installing npm@3 in the first place.

**Comment by Overdrivr** [comment]
Ok the issue is described here 
https://github.com/npm/npm/releases
Using the force flag, I managed to fall back to 2
```
npm install -f -g npm@2
```
So then I tried using windows sdk console
```
npm install -g strongloop
```
I have a bunch of modules that seem to fail to compile but... For the first time installation process completes ! :trophy: 
I can type 
```
slc -v
```
And that returns the following

```
C:\Program Files\Microsoft SDKs\Windows\v7.1>slc -v
strongloop v5.0.1 (node v4.1.0)
├── strong-arc@1.8.1 (15814c0)
├── strong-build@2.0.5 (70e7f5b)
├── strong-deploy@3.1.0 (72dbb7c)
├── strong-mesh-models@8.0.0 (1b220f7)
├── strong-pm@5.0.0 (dbaa1a5)
├── strong-registry@1.1.5 (f46e58f)
├── strong-start@1.3.1 (b17cbc0)
├─┬ strong-supervisor@3.2.1 (50fcfa5)
│ └── strong-agent@2.0.1 (8554cb3)
├── generator-loopback@1.12.0 (01a38d5)
├── node-inspector@0.7.4
└── nodefly-register@0.3.3

C:\Program Files\Microsoft SDKs\Windows\v7.1>
```
So I guess as long as the npm dependency issue is not fixed strongloop cannot be installed with npm3+.

Shall we close this ticket then ?
Again, thanks for your help ;)

**Comment by gustavomick** [comment]
i believe this is likned to same problem 

```
C:\nodejs\node_modules\strongloop\node_modules\node-inspector\node_modules\nan\nan.h(409): error C2039: 'ExternalAsciiS
tringResource': is not a member of 'v8::String' [C:\nodejs\node_modules\strongloop\node_modules\node-inspector\node_mod
ules\ws\build\bufferutil.vcxproj]
```

this failed with npm 2.x/3.x and node v4.1.1 on windows 10 ( fresh install with vs2015) . after that i installed slc with node 0.12.7 and was successful.

**Comment by surfjedi** [comment]
+1
